### ISSUE TITLE
Memory session cache eviction order is arbitrary

### ISSUE THREAD

**Issue by briansmith** [ctz/rustls#469]
@dhobsd wrote:
> ClientSessionMemoryCache iterates keys of a std::collections::HashMap to evict. This process executes in an arbitrary order, and it's possible to evict the just-inserted key. 

I noticed this too. When the cache is full, we evict an arbitrary element, instead of the least-recently used one.

**Comment by briansmith** [comment]
@dhobsd mentioned that Fuchsia has an implementation of an LRU memory cache already: https://fuchsia.googlesource.com/fuchsia/+/refs/changes/62/478362/24/src/lib/fuchsia-hyper/src/session_cache.rs.

Perhaps we can solve this by just having Google contribute their LRU implementation to replace the default memory cache implementation in Rustls?

**Comment by dhobsd** [comment]
Yeah, I'm happy to contribute -- given that it's part of what's needed for C.4 (but is a separate issue), I wonder if the desire is to implement it separately? We're stuck maintaining what we have in-tree anyway, so I'm happy to contribute however makes the most sense for rustls.

**Comment by briansmith** [comment]
The severity of this problem has decreased with PR #506, since even if the current session is removed from the cache during the handshake, it won't affect this handshake. Now this is merely a performance issue instead of a correctness issue, IIUC.

**Comment by dhobsd** [comment]
Yeah, agree. I think it's also worth mentioning that the correctness aspect isn't particularly security-weakening. Quick fix for this might just be to evict when full and we're inserting using the HashMap `entry` API. WDYT about something like this:

```diff
diff --git a/rustls/src/client/handy.rs b/rustls/src/client/handy.rs
index 712d27c..01fa62a 100644
--- a/rustls/src/client/handy.rs
+++ b/rustls/src/client/handy.rs
@@ -41,7 +41,7 @@ impl ClientSessionMemoryCache {
 
     fn limit_size(&self) {
         let mut cache = self.cache.lock().unwrap();
-        while cache.len() > self.max_entries {
+        while cache.len() >= self.max_entries {
             let k = cache.keys().next().unwrap().clone();
             cache.remove(&k);
         }
@@ -53,8 +53,12 @@ impl client::StoresClientSessions for ClientSessionMemoryCache {
         self.cache
             .lock()
             .unwrap()
-            .insert(key, value);
-        self.limit_size();
+            .entry(key)
+            .and_modify(|e| *e = value.clone())
+            .or_insert_with(|| {
+                self.limit_size();
+                value.clone()
+            });
         true
     }
```

**Comment by briansmith** [comment]
With that change, the eviction order is still arbitrary. We'd ensure that the new entry is always in the cache at the end, but we still wouldn't be having any prioritization of what gets evicted.

That said, I do think that's a very easy improvement over what is currently done, even if it doesn't fix this issue. Is it really guaranteed to be safe to call `self.limit_size()`, i.e. remove an entry, in `.or_insert_with()` though?

**Comment by dhobsd** [comment]
Ha! Yes, all good points. GitHub on Saturday mornings is not my best idea ever XD. I'm not sure how "unsafe" it is (Rust didn't hate it), but I imagine that at best it's a different spelling of the same behavior we have now.

Still being relatively new to Rust, I'm a bit surprised there aren't any standard linear collection types with efficient removal! Will poke around at adding an LRU on the keyspace soon.

**Comment by briansmith** [comment]
@dhobsd No worries. The Rust safety mechanism doesn't statically prevent deadlocks. If you run `cargo test test_clientsessionmemorycache` then some of those tests will fail (actually never complete) due to deadlock. The deadlock is caused by attempting to grab the lock inside `.or_insert_with()` while already holding it.

The other issue I see with that patch is that `value.clone()` is called twice. I think that's because the `and_modify` and `or_insert_with` combinators don't give you enough control over when the ownership of `value` is transfered. Check out draft PR #531 to see a way of avoiding the cloning. Though that has its own issues.

> Will poke around at adding an LRU on the keyspace soon.

I think that's a better way to move forward.

**Comment by dhobsd** [comment]
QQ -- work is work and rustls isn't always top priority -- are there rules on crate dependencies on rustls? In particular, an LRU is a complicated structure to implement in rust without use of unsafe, and I wonder whether that's something that folks want to maintain as part of this project (rather than using something like the [lru crate](https://docs.rs/lru/0.6.5/lru/)). I definitely underestimated the complexity of a doubly-linked list in this language. :|

**Comment by ctz** [comment]
Gah, I wrote a long reply here and github managed to lose it. So, briefly:

I have no fundamental objection to new crate dependencies, but a crate that implements collections using `unsafe` needs to be *very good*. eg, much evidence of ensuring bugs don't exist -- full MIRI coverage, many proptests, etc. clru looks a bit closer to this than lru.

I'm not sure the actual collection implementations belong in rustls, for similar reasons.

As a start to this, in f1285383 I've hoisted out that data structure underneath these two and done the naive thing -- as a starting point for further work. Notably, this only evicts based on original insertion order: not "use", and doesn't track updates to the same key. So this is better than it was, but still not fully there.
